// @rarible/sdk/build/sdk-blockchains/flow/test/sell-item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sellItem = void 0;
var tslib_1 = require("tslib");
var types_1 = require("@rarible/types");
var common_1 = require("./common");
function sellItem(sell, itemId, priceDecimals) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var submit, orderId;
        return (0, tslib_1.__generator)(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, sell.sell({
                        collectionId: common_1.testFlowCollection,
                    })];
                case 1:
                    submit = (_a.sent()).submit;
                    return [4 /*yield*/, submit({
                            amount: 1,
                            price: (0, types_1.toBigNumber)(priceDecimals),
                            currency: {
                                "@type": "FLOW_FT",
                                contract: common_1.testFlowToken,
                            },
                            itemId: itemId,
                        })];
                case 2:
                    orderId = _a.sent();
                    expect(orderId).toBeTruthy();
                    return [2 /*return*/, orderId];
            }
        });
    });
}
exports.sellItem = sellItem;
