// @rarible/sdk/build/sdk-blockchains/union/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUnionSdk = void 0;
var api_client_1 = require("@rarible/api-client");
var action_1 = require("@rarible/action");
var index_1 = require("../../index");
var middleware_1 = require("../../common/middleware/middleware");
function createUnionSdk(ethereum, flow, tezos, polygon) {
    return {
        balances: new UnionBalanceSdk({
            ETHEREUM: ethereum.balances,
            FLOW: flow.balances,
            TEZOS: tezos.balances,
            POLYGON: polygon.balances,
        }),
        nft: new UnionNftSdk({
            ETHEREUM: ethereum.nft,
            FLOW: flow.nft,
            TEZOS: tezos.nft,
            POLYGON: polygon.nft,
        }),
        order: new UnionOrderSdk({
            ETHEREUM: ethereum.order,
            FLOW: flow.order,
            TEZOS: tezos.order,
            POLYGON: polygon.order,
        }),
        restriction: new UnionRestrictionSdk({
            ETHEREUM: ethereum.restriction,
            FLOW: flow.restriction,
            TEZOS: tezos.restriction,
            POLYGON: polygon.restriction,
        }),
    };
}
exports.createUnionSdk = createUnionSdk;
var UnionOrderSdk = /** @class */ (function () {
    function UnionOrderSdk(instances) {
        var _this = this;
        this.instances = instances;
        this.cancel = action_1.Action.create({
            id: "send-tx",
            run: function (value) { return _this.instances[extractBlockchain(value.orderId)].cancel(value); },
        });
        this.bid = this.bid.bind(this);
        this.bidUpdate = this.bidUpdate.bind(this);
        this.fill = this.fill.bind(this);
        this.buy = this.buy.bind(this);
        this.acceptBid = this.acceptBid.bind(this);
        this.sell = this.sell.bind(this);
        this.sellUpdate = this.sellUpdate.bind(this);
    }
    UnionOrderSdk.prototype.bid = function (request) {
        return this.instances[extractBlockchain(getBidEntity(request))].bid(request);
    };
    UnionOrderSdk.prototype.bidUpdate = function (request) {
        return this.instances[extractBlockchain(request.orderId)].bidUpdate(request);
    };
    /**
     * @deprecated
     * @param request
     */
    UnionOrderSdk.prototype.fill = function (request) {
        return this.instances[extractBlockchain(getOrderId(request))].fill(request);
    };
    UnionOrderSdk.prototype.buy = function (request) {
        return this.instances[extractBlockchain(getOrderId(request))].buy(request);
    };
    UnionOrderSdk.prototype.acceptBid = function (request) {
        return this.instances[extractBlockchain(getOrderId(request))].acceptBid(request);
    };
    UnionOrderSdk.prototype.sell = function (request) {
        return this.instances[extractBlockchain(request.collectionId)].sell(request);
    };
    UnionOrderSdk.prototype.sellUpdate = function (request) {
        return this.instances[extractBlockchain(request.orderId)].sellUpdate(request);
    };
    return UnionOrderSdk;
}());
function getOrderId(req) {
    if ("orderId" in req) {
        return req.orderId;
    }
    else {
        return req.order.id;
    }
}
var UnionNftSdk = /** @class */ (function () {
    function UnionNftSdk(instances) {
        var _this = this;
        this.instances = instances;
        this.deploy = action_1.Action.create({
            id: "send-tx",
            run: function (request) { return _this.instances[request.blockchain].deploy(request); },
        });
        this.burn = this.burn.bind(this);
        this.mint = this.mint.bind(this);
        this.transfer = this.transfer.bind(this);
        this.preprocessMeta = middleware_1.Middlewarer.skipMiddleware(this.preprocessMeta.bind(this));
        this.generateTokenId = this.generateTokenId.bind(this);
    }
    UnionNftSdk.prototype.burn = function (request) {
        return this.instances[extractBlockchain(request.itemId)].burn(request);
    };
    UnionNftSdk.prototype.mint = function (request) {
        var collectionId = (0, index_1.getCollectionId)(request);
        return this.instances[extractBlockchain(collectionId)].mint(request);
    };
    UnionNftSdk.prototype.transfer = function (request) {
        return this.instances[extractBlockchain(request.itemId)].transfer(request);
    };
    UnionNftSdk.prototype.generateTokenId = function (prepare) {
        return this.instances[extractBlockchain(prepare.collection)].generateTokenId(prepare);
    };
    UnionNftSdk.prototype.preprocessMeta = function (request) {
        return this.instances[request.blockchain].preprocessMeta(request);
    };
    return UnionNftSdk;
}());
var UnionBalanceSdk = /** @class */ (function () {
    function UnionBalanceSdk(instances) {
        this.instances = instances;
        this.getBalance = this.getBalance.bind(this);
    }
    UnionBalanceSdk.prototype.getBalance = function (address, assetType) {
        return this.instances[extractBlockchain(address)].getBalance(address, assetType);
    };
    return UnionBalanceSdk;
}());
var UnionRestrictionSdk = /** @class */ (function () {
    function UnionRestrictionSdk(instances) {
        this.instances = instances;
    }
    UnionRestrictionSdk.prototype.canTransfer = function (itemId, from, to) {
        return this.instances[extractBlockchain(itemId)].canTransfer(itemId, from, to);
    };
    return UnionRestrictionSdk;
}());
var blockchains = [
    api_client_1.Blockchain.ETHEREUM,
    api_client_1.Blockchain.FLOW,
    api_client_1.Blockchain.TEZOS,
    api_client_1.Blockchain.POLYGON,
];
function extractBlockchain(value) {
    var idx = value.indexOf(":");
    if (idx === -1) {
        throw new Error("Unable to extract blockchain from " + value);
    }
    var start = value.substring(0, idx);
    for (var _i = 0, blockchains_1 = blockchains; _i < blockchains_1.length; _i++) {
        var blockchain = blockchains_1[_i];
        if (blockchain === start) {
            return blockchain;
        }
    }
    throw new Error("Unable to extract blockchain from " + value);
}
function getBidEntity(request) {
    if ("itemId" in request) {
        return request.itemId;
    }
    else if ("collectionId" in request) {
        return request.collectionId;
    }
    else {
        throw new Error("Bit request should contains itemId or collectionId");
    }
}
