// @rarible/tezos-sdk/dist/main/script.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("./index");
var in_memory_provider_1 = require("../providers/in_memory/in_memory_provider");
var yargs_1 = require("yargs");
var bignumber_js_1 = require("bignumber.js");
function main() {
    return __awaiter(this, void 0, void 0, function () {
        var argv, token_id_opt, token_id, royalties0, royalties, amount, metadata, tezos, config, provider, to, _a, owner, _b, fee_receiver, _c, asset_class, _d, op_transfer, op_mint, op_burn, op_deploy_fa2, op_deploy_royalties, op_token_metadata, op_metadata_uri, op_deploy_fill, op_deploy_exchange, op_deploy_transfer_proxy, op_deploy_transfer_manager, arg_update, op_update;
        return __generator(this, function (_e) {
            switch (_e.label) {
                case 0: return [4 /*yield*/, (0, yargs_1.default)(process.argv.slice(2)).options({
                        edsk: { type: 'string', default: 'edsk4CmgW9r4fwqtsT6x2bB7BdVcERxLPt6poFXGpk1gTKbqR43G5H' },
                        endpoint: { type: 'string', default: 'https://hangzhou.tz.functori.com' },
                        exchange: { type: 'string', default: 'KT1ULGjK8FtaJ9QqCgJVN14B6tY76Ykaz6M8' },
                        contract: { type: 'string', default: '' },
                        royalties_contract: { type: 'string', default: 'KT1WKRXswxEpTbVg2pGgofzwZCNKjAVcuMgh' },
                        token_id: { type: 'number' },
                        royalties: { type: 'string', default: '{}' },
                        amount: { type: 'number' },
                        metadata: { type: 'string', default: '{}' },
                        metadata_key: { type: 'string', default: '' },
                        metadata_value: { type: 'string', default: '' },
                        to: { type: 'string' },
                        owner: { type: 'string' },
                        receiver: { type: 'string' },
                        fee: { type: 'number', default: 0 },
                        operator: { type: 'string', default: '' },
                        fill: { type: 'string', default: 'KT1FAndThSQsVqYQVPHGSG5sQPk1XZycNBvL' },
                        transfer_proxy: { type: 'string', default: 'KT1Qypf9A7DHoAeesu5hj8v6iKwHsJb1RUR2' },
                        transfer_manager: { type: 'string', default: 'KT1DyDkW16XBuFzpLkXKraD46SAxQDrha5gm' },
                        fee_receiver: { type: 'string' },
                        protocol_fee: { type: 'number', default: 0 },
                        wrapper: { type: 'string', default: 'KT1LkKaeLBvTBo6knGeN5RsEunERCaqVcLr9' },
                    }).argv];
                case 1:
                    argv = _e.sent();
                    token_id_opt = (argv.token_id != undefined) ? new bignumber_js_1.default(argv.token_id) : undefined;
                    token_id = (argv.token_id != undefined) ? new bignumber_js_1.default(argv.token_id) : new bignumber_js_1.default(0);
                    royalties0 = JSON.parse(argv.royalties);
                    royalties = {};
                    if (royalties0) {
                        Object.keys(royalties0).forEach(function (k) {
                            royalties[k] = new bignumber_js_1.default(royalties0[k]);
                        });
                    }
                    amount = (argv.amount) ? new bignumber_js_1.default(argv.amount) : undefined;
                    metadata = JSON.parse(argv.metadata);
                    tezos = (0, in_memory_provider_1.in_memory_provider)(argv.edsk, argv.endpoint);
                    config = {
                        exchange: argv.exchange,
                        transfer_proxy: argv.transfer_proxy,
                        fees: new bignumber_js_1.default(argv.protocol_fee),
                        nft_public: "",
                        mt_public: "",
                        api: "https://localhost:8080/v0.1",
                        api_permit: "https://localhost:8081/v0.1",
                        permit_whitelist: [],
                        wrapper: argv.wrapper
                    };
                    provider = {
                        tezos: tezos,
                        config: config
                    };
                    if (!(argv.to)) return [3 /*break*/, 2];
                    _a = argv.to;
                    return [3 /*break*/, 4];
                case 2: return [4 /*yield*/, provider.tezos.address()];
                case 3:
                    _a = _e.sent();
                    _e.label = 4;
                case 4:
                    to = _a;
                    if (!(argv.owner)) return [3 /*break*/, 5];
                    _b = argv.owner;
                    return [3 /*break*/, 7];
                case 5: return [4 /*yield*/, provider.tezos.address()];
                case 6:
                    _b = _e.sent();
                    _e.label = 7;
                case 7:
                    owner = _b;
                    if (!(argv.fee_receiver)) return [3 /*break*/, 8];
                    _c = argv.fee_receiver;
                    return [3 /*break*/, 10];
                case 8: return [4 /*yield*/, provider.tezos.address()];
                case 9:
                    _c = _e.sent();
                    _e.label = 10;
                case 10:
                    fee_receiver = _c;
                    asset_class = (amount == undefined) ? "NFT" : "MT";
                    _d = argv._[0];
                    switch (_d) {
                        case 'transfer': return [3 /*break*/, 11];
                        case 'mint': return [3 /*break*/, 14];
                        case 'burn': return [3 /*break*/, 17];
                        case 'deploy_nft': return [3 /*break*/, 20];
                        case 'deploy_royalties': return [3 /*break*/, 23];
                        case 'set_token_metadata': return [3 /*break*/, 26];
                        case 'set_metadata': return [3 /*break*/, 29];
                        case 'deploy_fill': return [3 /*break*/, 32];
                        case 'deploy_exchange': return [3 /*break*/, 35];
                        case 'deploy_transfer_proxy': return [3 /*break*/, 38];
                        case 'deploy_transfer_manager': return [3 /*break*/, 41];
                        case 'update_operators_for_all': return [3 /*break*/, 44];
                    }
                    return [3 /*break*/, 47];
                case 11:
                    console.log("transfer");
                    return [4 /*yield*/, (0, index_1.transfer)(provider, { asset_class: asset_class, contract: argv.contract, token_id: token_id }, to, amount)];
                case 12:
                    op_transfer = _e.sent();
                    return [4 /*yield*/, op_transfer.confirmation()];
                case 13:
                    _e.sent();
                    console.log(op_transfer.hash);
                    return [3 /*break*/, 47];
                case 14:
                    console.log("mint");
                    return [4 /*yield*/, (0, index_1.mint)(provider, argv.contract, royalties, amount, token_id_opt, metadata, argv.owner)];
                case 15:
                    op_mint = _e.sent();
                    return [4 /*yield*/, op_mint.confirmation()];
                case 16:
                    _e.sent();
                    console.log(op_mint.hash);
                    return [3 /*break*/, 47];
                case 17:
                    console.log("burn");
                    return [4 /*yield*/, (0, index_1.burn)(provider, { asset_class: asset_class, contract: argv.contract, token_id: token_id }, amount)];
                case 18:
                    op_burn = _e.sent();
                    return [4 /*yield*/, op_burn.confirmation()];
                case 19:
                    _e.sent();
                    console.log(op_burn.hash);
                    return [3 /*break*/, 47];
                case 20:
                    console.log("deploy nft");
                    return [4 /*yield*/, (0, index_1.deploy_nft_public)(provider, owner)];
                case 21:
                    op_deploy_fa2 = _e.sent();
                    return [4 /*yield*/, op_deploy_fa2.confirmation()];
                case 22:
                    _e.sent();
                    console.log(op_deploy_fa2.contract);
                    return [3 /*break*/, 47];
                case 23:
                    console.log("deploy royalties");
                    return [4 /*yield*/, (0, index_1.deploy_royalties)(provider, owner)];
                case 24:
                    op_deploy_royalties = _e.sent();
                    return [4 /*yield*/, op_deploy_royalties.confirmation()];
                case 25:
                    _e.sent();
                    console.log(op_deploy_royalties.contract);
                    return [3 /*break*/, 47];
                case 26:
                    console.log("set token metadata");
                    return [4 /*yield*/, (0, index_1.set_token_metadata)(provider, argv.contract, token_id, metadata)];
                case 27:
                    op_token_metadata = _e.sent();
                    return [4 /*yield*/, op_token_metadata.confirmation()];
                case 28:
                    _e.sent();
                    console.log(op_token_metadata.hash);
                    return [3 /*break*/, 47];
                case 29:
                    console.log("set metadata uri");
                    return [4 /*yield*/, (0, index_1.set_metadata)(provider, argv.contract, argv.metadata_key, argv.metadata_value)];
                case 30:
                    op_metadata_uri = _e.sent();
                    return [4 /*yield*/, op_metadata_uri.confirmation()];
                case 31:
                    _e.sent();
                    console.log(op_metadata_uri.hash);
                    return [3 /*break*/, 47];
                case 32:
                    console.log("deploy fill");
                    return [4 /*yield*/, (0, index_1.deploy_fill)(provider, owner)];
                case 33:
                    op_deploy_fill = _e.sent();
                    return [4 /*yield*/, op_deploy_fill.confirmation()];
                case 34:
                    _e.sent();
                    console.log(op_deploy_fill.contract);
                    return [3 /*break*/, 47];
                case 35:
                    console.log("deploy exchange");
                    return [4 /*yield*/, (0, index_1.deploy_exchange)(provider, owner, argv.transfer_manager, argv.royalties_contract, argv.fill)];
                case 36:
                    op_deploy_exchange = _e.sent();
                    return [4 /*yield*/, op_deploy_exchange.confirmation()];
                case 37:
                    _e.sent();
                    console.log(op_deploy_exchange.contract);
                    return [3 /*break*/, 47];
                case 38:
                    console.log("deploy transfer proxy");
                    return [4 /*yield*/, (0, index_1.deploy_transfer_proxy)(provider, owner)];
                case 39:
                    op_deploy_transfer_proxy = _e.sent();
                    return [4 /*yield*/, op_deploy_transfer_proxy.confirmation()];
                case 40:
                    _e.sent();
                    console.log(op_deploy_transfer_proxy.contract);
                    return [3 /*break*/, 47];
                case 41:
                    console.log("deploy transfer manager");
                    return [4 /*yield*/, (0, index_1.deploy_transfer_manager)(provider, owner, fee_receiver, new bignumber_js_1.default(argv.protocol_fee))];
                case 42:
                    op_deploy_transfer_manager = _e.sent();
                    return [4 /*yield*/, op_deploy_transfer_manager.confirmation()];
                case 43:
                    _e.sent();
                    console.log(op_deploy_transfer_manager.contract);
                    return [3 /*break*/, 47];
                case 44:
                    console.log('update operators for all');
                    arg_update = {
                        destination: argv.contract,
                        entrypoint: "update_operators_for_all",
                        parameter: [{ prim: 'Left', args: [{ string: argv.operator }] }]
                    };
                    return [4 /*yield*/, (0, index_1.send)(provider, arg_update)];
                case 45:
                    op_update = _e.sent();
                    return [4 /*yield*/, op_update.confirmation()];
                case 46:
                    _e.sent();
                    console.log(op_update.hash);
                    return [3 /*break*/, 47];
                case 47: return [2 /*return*/];
            }
        });
    });
}
main();
