// @rarible/sdk/build/sdk-blockchains/tezos/sell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TezosSell = void 0;
var tslib_1 = require("tslib");
var sell_1 = require("@rarible/tezos-sdk/dist/order/sell");
// eslint-disable-next-line camelcase
var tezos_sdk_1 = require("@rarible/tezos-sdk");
var action_1 = require("@rarible/action");
var bignumber_js_1 = (0, tslib_1.__importDefault)(require("bignumber.js"));
var api_client_1 = require("@rarible/api-client");
var domain_1 = require("../../types/order/fill/domain");
var retry_1 = require("../../common/retry");
var common_1 = require("./common");
var TezosSell = /** @class */ (function () {
    function TezosSell(provider, apis) {
        this.provider = provider;
        this.apis = apis;
        this.sell = this.sell.bind(this);
        this.update = this.update.bind(this);
    }
    TezosSell.prototype.parseTakeAssetType = function (type) {
        switch (type["@type"]) {
            case "XTZ":
                return {
                    asset_class: type["@type"],
                };
            case "TEZOS_FT":
                return {
                    asset_class: "FT",
                    contract: (0, common_1.convertFromContractAddress)(type.contract),
                    token_id: type.tokenId ? new bignumber_js_1.default(type.tokenId) : undefined,
                };
            default: {
                throw new Error("Unsupported take asset type");
            }
        }
    };
    TezosSell.prototype.sell = function (prepareRequest) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a, domain, contract, itemCollection, submit;
            var _this = this;
            return (0, tslib_1.__generator)(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = prepareRequest.collectionId.split(":"), domain = _a[0], contract = _a[1];
                        if (domain !== api_client_1.Blockchain.TEZOS) {
                            throw new Error("Not an tezos item");
                        }
                        return [4 /*yield*/, this.apis.collection.getNftCollectionById({
                                collection: contract,
                            })];
                    case 1:
                        itemCollection = _b.sent();
                        submit = action_1.Action.create({
                            id: "send-tx",
                            run: function (request) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                                var provider, makerPublicKey, itemId, item, tezosRequest, sellOrder;
                                var _a;
                                var _this = this;
                                return (0, tslib_1.__generator)(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            provider = (0, common_1.getRequiredProvider)(this.provider);
                                            return [4 /*yield*/, (0, common_1.getMakerPublicKey)(provider)];
                                        case 1:
                                            makerPublicKey = _b.sent();
                                            itemId = (0, common_1.getTezosItemData)(request.itemId).itemId;
                                            return [4 /*yield*/, (0, retry_1.retry)(90, 1000, function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                                                    return (0, tslib_1.__generator)(this, function (_a) {
                                                        return [2 /*return*/, this.apis.item.getNftItemById({ itemId: itemId })];
                                                    });
                                                }); })];
                                        case 2:
                                            item = _b.sent();
                                            _a = {
                                                maker: (0, tezos_sdk_1.pk_to_pkh)(makerPublicKey),
                                                maker_edpk: makerPublicKey,
                                                make_asset_type: {
                                                    asset_class: itemCollection.type,
                                                    contract: item.contract,
                                                    token_id: new bignumber_js_1.default(item.tokenId),
                                                },
                                                take_asset_type: this.parseTakeAssetType(request.currency),
                                                amount: new bignumber_js_1.default(request.amount),
                                                price: new bignumber_js_1.default(request.price)
                                            };
                                            return [4 /*yield*/, (0, common_1.getPayouts)(provider, request.payouts)];
                                        case 3:
                                            tezosRequest = (_a.payouts = _b.sent(),
                                                _a.origin_fees = (0, common_1.convertOrderPayout)(request.originFees),
                                                _a);
                                            return [4 /*yield*/, (0, sell_1.sell)(provider, tezosRequest)];
                                        case 4:
                                            sellOrder = _b.sent();
                                            return [2 /*return*/, (0, common_1.convertTezosOrderId)(sellOrder.hash)];
                                    }
                                });
                            }); },
                        });
                        return [2 /*return*/, {
                                multiple: itemCollection.type === "MT",
                                originFeeSupport: domain_1.OriginFeeSupport.FULL,
                                payoutsSupport: domain_1.PayoutsSupport.MULTIPLE,
                                supportedCurrencies: (0, common_1.getSupportedCurrencies)(),
                                baseFee: parseInt(this.provider.config.fees.toString()),
                                submit: submit,
                            }];
                }
            });
        });
    };
    TezosSell.prototype.update = function (request) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var orderId, order, updateAction;
            var _this = this;
            return (0, tslib_1.__generator)(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        orderId = (0, common_1.getTezosOrderId)(request.orderId);
                        return [4 /*yield*/, this.apis.order.getOrderByHash({ hash: orderId })];
                    case 1:
                        order = _a.sent();
                        if (!order) {
                            throw new Error("Order has not been found");
                        }
                        updateAction = action_1.Action.create({
                            id: "send-tx",
                            run: function (updateRequest) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                                var provider, orderForm, updatedOrder;
                                var _a, _b;
                                return (0, tslib_1.__generator)(this, function (_c) {
                                    switch (_c.label) {
                                        case 0:
                                            provider = (0, common_1.getRequiredProvider)(this.provider);
                                            orderForm = {
                                                type: "RARIBLE_V2",
                                                maker: order.maker,
                                                maker_edpk: order.makerEdpk,
                                                taker_edpk: order.takerEdpk,
                                                make: (0, common_1.covertToLibAsset)(order.make),
                                                take: (0, tslib_1.__assign)((0, tslib_1.__assign)({}, (0, common_1.covertToLibAsset)(order.take)), { value: new bignumber_js_1.default(updateRequest.price).multipliedBy(order.make.value) }),
                                                salt: order.salt,
                                                start: order.start,
                                                end: order.end,
                                                signature: order.signature,
                                                data: {
                                                    data_type: "V1",
                                                    payouts: ((_a = order.data.payouts) === null || _a === void 0 ? void 0 : _a.map(function (p) { return ({
                                                        account: p.account,
                                                        value: new bignumber_js_1.default(p.value),
                                                    }); })) || [],
                                                    origin_fees: ((_b = order.data.originFees) === null || _b === void 0 ? void 0 : _b.map(function (p) { return ({
                                                        account: p.account,
                                                        value: new bignumber_js_1.default(p.value),
                                                    }); })) || [],
                                                },
                                            };
                                            return [4 /*yield*/, (0, tezos_sdk_1.upsert_order)(provider, orderForm, true, true)];
                                        case 1:
                                            updatedOrder = _c.sent();
                                            return [2 /*return*/, (0, common_1.convertTezosOrderId)(updatedOrder.hash)];
                                    }
                                });
                            }); },
                        });
                        return [2 /*return*/, {
                                originFeeSupport: domain_1.OriginFeeSupport.FULL,
                                payoutsSupport: domain_1.PayoutsSupport.MULTIPLE,
                                supportedCurrencies: (0, common_1.getSupportedCurrencies)(),
                                baseFee: parseInt(this.provider.config.fees.toString()),
                                submit: updateAction,
                            }];
                }
            });
        });
    };
    return TezosSell;
}());
exports.TezosSell = TezosSell;
