// @rarible/sdk/build/common/logger/logger-middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInternalLoggerMiddleware = void 0;
var tslib_1 = require("tslib");
var build_1 = require("@rarible/logger/build");
var api_client_1 = require("@rarible/api-client");
var axios_1 = (0, tslib_1.__importDefault)(require("axios"));
var domain_1 = require("../../domain");
var loggerConfig = {
    service: "union-sdk",
    elkUrl: "https://logging.rarible.com/",
};
function getWalletInfo(wallet) {
    return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
        var info, _a;
        return (0, tslib_1.__generator)(this, function (_b) {
            switch (_b.label) {
                case 0:
                    info = {
                        "wallet.blockchain": wallet.blockchain,
                    };
                    _a = wallet.blockchain;
                    switch (_a) {
                        case api_client_1.Blockchain.ETHEREUM: return [3 /*break*/, 1];
                        case api_client_1.Blockchain.FLOW: return [3 /*break*/, 3];
                        case api_client_1.Blockchain.TEZOS: return [3 /*break*/, 5];
                    }
                    return [3 /*break*/, 7];
                case 1: return [4 /*yield*/, Promise.all([wallet.ethereum.getChainId(), wallet.ethereum.getFrom()])
                        .then(function (_a) {
                        var chainId = _a[0], address = _a[1];
                        info["wallet.address"] = address;
                        info["wallet.chainId"] = chainId;
                    })
                        .catch(function (err) {
                        info["wallet.address"] = "unknown (" + (err && err.toString()) + ")";
                    })];
                case 2:
                    _b.sent();
                    return [3 /*break*/, 8];
                case 3: return [4 /*yield*/, wallet.fcl.currentUser().snapshot()
                        .then(function (userData) {
                        info["wallet.address"] = userData.addr;
                        info["wallet.chainId"] = userData.cid;
                    })
                        .catch(function (err) {
                        info["wallet.address"] = "unknown (" + (err && err.toString()) + ")";
                    })];
                case 4:
                    _b.sent();
                    return [3 /*break*/, 8];
                case 5:
                    info["wallet.tezos.kind"] = wallet.provider.kind;
                    return [4 /*yield*/, Promise.all([wallet.provider.chain_id(), wallet.provider.address()])
                            .then(function (_a) {
                            var chainId = _a[0], address = _a[1];
                            info["wallet.address"] = address;
                            info["wallet.chainId"] = chainId;
                        })
                            .catch(function (err) {
                            info["wallet.address"] = "unknown (" + (err && err.toString()) + ")";
                        })];
                case 6:
                    _b.sent();
                    return [3 /*break*/, 8];
                case 7:
                    info["wallet.address"] = "unknown";
                    _b.label = 8;
                case 8: return [2 /*return*/, info];
            }
        });
    });
}
function getInternalLoggerMiddleware(logsLevel, sdkContext) {
    var _this = this;
    var getContext = function () { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
        var _a, _b;
        return (0, tslib_1.__generator)(this, function (_c) {
            switch (_c.label) {
                case 0:
                    _a = [{ service: loggerConfig.service, environment: sdkContext.env }];
                    if (!sdkContext.wallet) return [3 /*break*/, 2];
                    return [4 /*yield*/, getWalletInfo(sdkContext.wallet)];
                case 1:
                    _b = _c.sent();
                    return [3 /*break*/, 3];
                case 2:
                    _b = {};
                    _c.label = 3;
                case 3: return [2 /*return*/, tslib_1.__assign.apply(void 0, _a.concat([(_b)]))];
            }
        });
    }); };
    var remoteLogger = new build_1.RemoteLogger(function (msg) { return axios_1.default.post(loggerConfig.elkUrl, msg); }, {
        initialContext: getContext(),
    });
    return function (callable, args) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
        var time;
        var _this = this;
        return (0, tslib_1.__generator)(this, function (_a) {
            time = Date.now();
            return [2 /*return*/, [callable, function (responsePromise) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                        var res, err_1;
                        return (0, tslib_1.__generator)(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    _a.trys.push([0, 2, , 3]);
                                    return [4 /*yield*/, responsePromise];
                                case 1:
                                    res = _a.sent();
                                    if (logsLevel >= domain_1.LogsLevel.TRACE) {
                                        remoteLogger.trace(callable.name, {
                                            time: (Date.now() - time) / 1000,
                                            args: args,
                                            response: res,
                                        });
                                    }
                                    return [3 /*break*/, 3];
                                case 2:
                                    err_1 = _a.sent();
                                    if (logsLevel >= domain_1.LogsLevel.ERROR) {
                                        remoteLogger.error(callable.name, {
                                            time: (Date.now() - time) / 1000,
                                            error: err_1.toString(),
                                            args: args,
                                        });
                                    }
                                    return [3 /*break*/, 3];
                                case 3: return [2 /*return*/, responsePromise];
                            }
                        });
                    }); }]];
        });
    }); };
}
exports.getInternalLoggerMiddleware = getInternalLoggerMiddleware;
