// @rarible/sdk/build/sdk-blockchains/ethereum/sell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EthereumSell = void 0;
var tslib_1 = require("tslib");
var types_1 = require("@rarible/types");
var domain_1 = require("../../types/order/fill/domain");
var common = (0, tslib_1.__importStar)(require("./common"));
var common_1 = require("./common");
var EthereumSell = /** @class */ (function () {
    function EthereumSell(sdk, network) {
        this.sdk = sdk;
        this.blockchain = (0, common_1.getEVMBlockchain)(network);
        this.sell = this.sell.bind(this);
        this.update = this.update.bind(this);
    }
    EthereumSell.prototype.sell = function (request) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a, domain, contract, collection, sellAction;
            var _b;
            var _this = this;
            return (0, tslib_1.__generator)(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = request.collectionId.split(":"), domain = _a[0], contract = _a[1];
                        if (!(0, common_1.isEVMBlockchain)(domain)) {
                            throw new Error("Not an ethereum item");
                        }
                        return [4 /*yield*/, this.sdk.apis.nftCollection.getNftCollectionById({
                                collection: contract,
                            })];
                    case 1:
                        collection = _c.sent();
                        sellAction = this.sdk.order.sell
                            .before(function (sellFormRequest) { return (0, tslib_1.__awaiter)(_this, void 0, void 0, function () {
                            var itemId, item;
                            return (0, tslib_1.__generator)(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        itemId = (0, common_1.getEthereumItemId)(sellFormRequest.itemId).itemId;
                                        return [4 /*yield*/, this.sdk.apis.nftItem.getNftItemById({ itemId: itemId })];
                                    case 1:
                                        item = _a.sent();
                                        return [2 /*return*/, {
                                                makeAssetType: {
                                                    tokenId: item.tokenId,
                                                    contract: item.contract,
                                                },
                                                amount: sellFormRequest.amount,
                                                takeAssetType: common.getEthTakeAssetType(sellFormRequest.currency),
                                                priceDecimal: sellFormRequest.price,
                                                payouts: common.toEthereumParts(sellFormRequest.payouts),
                                                originFees: common.toEthereumParts(sellFormRequest.originFees),
                                            }];
                                }
                            });
                        }); })
                            .after(function (order) { return common.convertEthereumOrderHash(order.hash, _this.blockchain); });
                        _b = {
                            originFeeSupport: domain_1.OriginFeeSupport.FULL,
                            payoutsSupport: domain_1.PayoutsSupport.MULTIPLE,
                            multiple: collection.type === "ERC1155",
                            supportedCurrencies: common.getSupportedCurrencies()
                        };
                        return [4 /*yield*/, this.sdk.order.getBaseOrderFee()];
                    case 2: return [2 /*return*/, (_b.baseFee = _c.sent(),
                            _b.submit = sellAction,
                            _b)];
                }
            });
        });
    };
    EthereumSell.prototype.update = function (prepareRequest) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _a, blockchain, hash, order, sellUpdateAction;
            var _b;
            var _this = this;
            return (0, tslib_1.__generator)(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!prepareRequest.orderId) {
                            throw new Error("OrderId has not been specified");
                        }
                        _a = prepareRequest.orderId.split(":"), blockchain = _a[0], hash = _a[1];
                        if (!(0, common_1.isEVMBlockchain)(blockchain)) {
                            throw new Error("Not an ethereum order");
                        }
                        return [4 /*yield*/, this.sdk.apis.order.getOrderByHash({ hash: hash })];
                    case 1:
                        order = _c.sent();
                        if (order.type !== "RARIBLE_V2" && order.type !== "RARIBLE_V1") {
                            throw new Error("Unable to update sell " + JSON.stringify(order));
                        }
                        sellUpdateAction = this.sdk.order.sellUpdate
                            .before(function (request) { return ({
                            orderHash: (0, types_1.toWord)(hash),
                            priceDecimal: request.price,
                        }); })
                            .after(function (order) { return common.convertEthereumOrderHash(order.hash, _this.blockchain); });
                        _b = {
                            originFeeSupport: order.type === "RARIBLE_V2" ? domain_1.OriginFeeSupport.FULL : domain_1.OriginFeeSupport.AMOUNT_ONLY,
                            payoutsSupport: order.type === "RARIBLE_V2" ? domain_1.PayoutsSupport.MULTIPLE : domain_1.PayoutsSupport.SINGLE,
                            supportedCurrencies: common.getSupportedCurrencies()
                        };
                        return [4 /*yield*/, this.sdk.order.getBaseOrderFee(order.type)];
                    case 2: return [2 /*return*/, (_b.baseFee = _c.sent(),
                            _b.submit = sellUpdateAction,
                            _b)];
                }
            });
        });
    };
    return EthereumSell;
}());
exports.EthereumSell = EthereumSell;
