// @rarible/sdk/build/types/nft/mint/prepare-mint-request.type.validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isHasCollectionId = exports.validateHasCollectionId = exports.isHasCollection = exports.validateHasCollection = exports.isPrepareMintRequest = exports.validatePrepareMintRequest = exports.SCHEMA = void 0;
var tslib_1 = require("tslib");
// @ts-nocheck
// eslint-disable
// This file is generated by create-validator-ts
var ajv_1 = (0, tslib_1.__importDefault)(require("ajv"));
exports.SCHEMA = {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "definitions": {
        "PrepareMintRequest": {
            "anyOf": [
                {
                    "type": "object",
                    "additionalProperties": false,
                    "properties": {
                        "collection": {
                            "$ref": "#/definitions/Collection"
                        },
                        "tokenId": {
                            "$ref": "#/definitions/TokenId"
                        }
                    },
                    "required": [
                        "collection"
                    ]
                },
                {
                    "type": "object",
                    "additionalProperties": false,
                    "properties": {
                        "collectionId": {
                            "$ref": "#/definitions/ContractAddress"
                        },
                        "tokenId": {
                            "$ref": "#/definitions/TokenId"
                        }
                    },
                    "required": [
                        "collectionId"
                    ]
                }
            ]
        },
        "TokenId": {
            "type": "object",
            "properties": {
                "tokenId": {
                    "type": "string"
                },
                "signature": {
                    "type": "object",
                    "properties": {
                        "v": {
                            "type": "number"
                        },
                        "r": {
                            "$ref": "#/definitions/Binary"
                        },
                        "s": {
                            "$ref": "#/definitions/Binary"
                        }
                    },
                    "required": [
                        "v",
                        "r",
                        "s"
                    ],
                    "additionalProperties": false
                }
            },
            "required": [
                "tokenId",
                "signature"
            ],
            "additionalProperties": false
        },
        "Binary": {
            "type": "string"
        },
        "Collection": {
            "type": "object",
            "properties": {
                "id": {
                    "$ref": "#/definitions/ContractAddress"
                },
                "blockchain": {
                    "$ref": "#/definitions/Blockchain"
                },
                "type": {
                    "$ref": "#/definitions/CollectionType"
                },
                "name": {
                    "type": "string"
                },
                "symbol": {
                    "type": "string"
                },
                "owner": {
                    "$ref": "#/definitions/UnionAddress"
                },
                "features": {
                    "type": "array",
                    "items": {
                        "$ref": "#/definitions/CollectionFeatures"
                    }
                }
            },
            "required": [
                "id",
                "blockchain",
                "type",
                "name",
                "features"
            ],
            "additionalProperties": false
        },
        "ContractAddress": {
            "type": "string"
        },
        "Blockchain": {
            "type": "string",
            "enum": [
                "ETHEREUM",
                "POLYGON",
                "FLOW",
                "TEZOS"
            ]
        },
        "CollectionType": {
            "type": "string",
            "enum": [
                "CRYPTO_PUNKS",
                "ERC721",
                "ERC1155",
                "FLOW",
                "TEZOS_NFT",
                "TEZOS_MT"
            ]
        },
        "UnionAddress": {
            "type": "string"
        },
        "CollectionFeatures": {
            "type": "string",
            "enum": [
                "APPROVE_FOR_ALL",
                "SET_URI_PREFIX",
                "BURN",
                "MINT_WITH_ADDRESS",
                "SECONDARY_SALE_FEES",
                "MINT_AND_TRANSFER"
            ]
        },
        "HasCollection": {
            "type": "object",
            "properties": {
                "collection": {
                    "$ref": "#/definitions/Collection"
                }
            },
            "required": [
                "collection"
            ],
            "additionalProperties": false
        },
        "HasCollectionId": {
            "type": "object",
            "properties": {
                "collectionId": {
                    "$ref": "#/definitions/ContractAddress"
                }
            },
            "required": [
                "collectionId"
            ],
            "additionalProperties": false
        }
    }
};
var ajv = new ajv_1.default({ removeAdditional: true }).addSchema(exports.SCHEMA, "SCHEMA");
function validatePrepareMintRequest(payload) {
    if (!isPrepareMintRequest(payload)) {
        var error = new Error('invalid payload: PrepareMintRequest');
        error.name = "ValidationError";
        throw error;
    }
    return payload;
}
exports.validatePrepareMintRequest = validatePrepareMintRequest;
function isPrepareMintRequest(payload) {
    /** Schema is defined in {@link SCHEMA.definitions.PrepareMintRequest } **/
    var ajvValidate = ajv.compile({ "$ref": "SCHEMA#/definitions/PrepareMintRequest" });
    return ajvValidate(payload);
}
exports.isPrepareMintRequest = isPrepareMintRequest;
function validateHasCollection(payload) {
    if (!isHasCollection(payload)) {
        var error = new Error('invalid payload: HasCollection');
        error.name = "ValidationError";
        throw error;
    }
    return payload;
}
exports.validateHasCollection = validateHasCollection;
function isHasCollection(payload) {
    /** Schema is defined in {@link SCHEMA.definitions.HasCollection } **/
    var ajvValidate = ajv.compile({ "$ref": "SCHEMA#/definitions/HasCollection" });
    return ajvValidate(payload);
}
exports.isHasCollection = isHasCollection;
function validateHasCollectionId(payload) {
    if (!isHasCollectionId(payload)) {
        var error = new Error('invalid payload: HasCollectionId');
        error.name = "ValidationError";
        throw error;
    }
    return payload;
}
exports.validateHasCollectionId = validateHasCollectionId;
function isHasCollectionId(payload) {
    /** Schema is defined in {@link SCHEMA.definitions.HasCollectionId } **/
    var ajvValidate = ajv.compile({ "$ref": "SCHEMA#/definitions/HasCollectionId" });
    return ajvValidate(payload);
}
exports.isHasCollectionId = isHasCollectionId;
